wp-includes/components/class-wp-service-worker-uploaded-image-caching-component.php

```diff
@@ -0,0 +1,86 @@
+<?php
+/**
+ * WP_Service_Worker_Uploaded_Image_Caching_Component class.
+ *
+ * @package PWA
+ */
+
+/**
+ * Use a cache-first caching strategy for uploaded images.
+ *
+ * @since 0.6
+ */
+class WP_Service_Worker_Uploaded_Image_Caching_Component implements WP_Service_Worker_Component {
+
+	/**
+	 * Cache name.
+	 *
+	 * @var string
+	 */
+	const CACHE_NAME = 'uploaded-images';
+
+	/**
+	 * Adds the component functionality to the service worker.
+	 *
+	 * @since 0.6
+	 *
+	 * @param WP_Service_Worker_Scripts $scripts Instance to register service worker behavior with.
+	 */
+	public function serve( WP_Service_Worker_Scripts $scripts ) {
+		if ( is_admin() ) {
+			return;
+		}
+		$types = wp_get_ext_types();
+		if ( empty( $types['image'] ) ) {
+			return;
+		}
+
+		$upload_dir       = wp_get_upload_dir();
+		$image_extensions = $types['image'];
+
+		/*
+		 * Note that the path alone is used because CDN plugins may load from another domain. For example, given an
+		 * uploaded image located at:
+		 *   https://example.com/wp-content/uploads/2020/04/foo.png
+		 * Jetpack can change rewrite the URL to be:
+		 *   https://i2.wp.com/example.com/wp-content/uploads/2020/04/foo.png?fit=900%2C832&ssl=1
+		 * Therefore, the following will include any URL ending in an image file extension which also is also
+		 * preceded by '/wp-content/uploads/'.
+		 */
```

Comment:
This is not going to work. Specifically, when a CacheFirst strategy is used, if the third-party domain is not responding with CORS headers then Workbox will refuse to cache the opaque responses by default: https://developers.google.com/web/tools/workbox/guides/handle-third-party-requests

We should probably just assume that an image CDN will already have the assets cached with a far-future expires header anyway and not worry about explicitly caching the responses by the service worker beyond the browser cache.

Reply:
Additionally, it would be good to determine if WP sites generally have correct `Cache-Control` response headers being served for uploaded files. If so, and browsers are caching them for a long time, then it would not be necessary to add caching for uploaded images